Send metrics to statsD

benjben opened the pull request:
5 metrics exported for each load:
- `latency_collector_to_load_min`: delay between maximum collector timestamp and load timestamp
- `latency_collector_to_load_max`: delay between minimum collector timestamp and load timestamp
- `latency_shredder_start_to_load`: delay between start of shredder and load timestamp
- `latency_shredder_end_to_load`: delay between end of shredder and load timestamp
- `count_good`: number of good events loaded

benjben commented:
I'm putting `sendMetrics` into `Logging` so that `load` doesn't need to receive more arguments than before and also because I think that this is where it should live.

benjben commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/loading/Load.scala:
```diff
@@ -119,14 +120,9 @@ object Load {
+    loaded: LoaderMessage.ShreddingComplete
+  ): F[Unit] =
+    Logging[F].info(s"Folder ${loaded.base} loaded successfully") >>
+    Logging[F].sendMetrics(loaded)
```
@chuwy can you please confirm that if something goes wrong inside `sendMetrics`, it won't affect the result of `load` because we are using `<*` in front of `congratulate` ?

chuwy replied:
Yes, can confirm that. `<*` doesn't change order/semantics of execution, it just tells that RHS value will be ignored and LHS returned. Whereas `*>` does it vice versa - LHS ignored, RHS returned.

benjben replied:
That's what I had understood, perfect

chuwy commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/loading/Load.scala:
```diff
@@ -119,14 +120,9 @@ object Load {
+  private def congratulate[F[_]: Logging: Monad](
+    loaded: LoaderMessage.ShreddingComplete
+  ): F[Unit] =
+    Logging[F].info(s"Folder ${loaded.base} loaded successfully") >>
```
I thought we agreed to keep logging in place?

benjben replied:
Yes. I didn't want to compute the metrics twice, so I moved the logging [here](https://github.com/snowplow/snowplow-rdb-loader/pull/391/files#diff-5aefeab53daac289abca1f1ba409202804160cd9fa8996d3016016b86cf302f3R110).

benjben commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/dsl/Logging.scala:
```diff
@@ -100,6 +103,13 @@ object Logging {
+        for {
+          metrics <- statsD.sendMetrics(loaded)
+          metricsStr = metrics.map { case StatsD.KVMetric(k, v) => s"$k:$v"}.mkString(", ")
+          _ <- if(metricsStr.nonEmpty) info(s"Metrics: $metricsStr") else Sync[F].unit
```
Now that we have the metric with load latency this can never be empty, I need to remove the check

chuwy commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/dsl/StatsD.scala:
```diff
@@ -0,0 +1,92 @@
+  private def mkSocket[F[_]: ContextShift: Sync](blocker: Blocker): Resource[F, DatagramSocket] =
+    Resource.fromAutoCloseableBlocking(blocker)(Sync[F].delay(new DatagramSocket))
+
+  private def getMetrics[F[_]: Monad: Timer](loaded: LoaderMessage.ShreddingComplete): F[List[KVMetric]] =
```
`Functor: Clock` should be enough.

chuwy commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/dsl/StatsD.scala:
```diff
@@ -0,0 +1,92 @@
+                formatted = metrics.map(statsDFormat(config))
+                ip <- blocker.delay(InetAddress.getByName(config.hostname))
+                _ <- formatted.traverse_(sendMetric[F](blocker, socket, ip, config.port))
+              } yield (metrics)
```
```suggestion
              } yield metrics
```

chuwy commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/loading/Load.scala:
```diff
@@ -119,14 +120,9 @@ object Load {
+  private def congratulate[F[_]: Logging: Monad](
+    loaded: LoaderMessage.ShreddingComplete
+  ): F[Unit] =
+    Logging[F].info(s"Folder ${loaded.base} loaded successfully") >>
```
Just to re-iterate on what I said this morning - I think these logs were important not because they were showing numbers, but because they were human-readable and giving operator an understanding on what is going on in the pipeline. Just showing machine-readable strings would be the same.

Also, if we added something in 1.0.0 we should be careful about removing/changing it, unless we identified a problem with that format.

chuwy commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/dsl/Logging.scala:
```diff
@@ -54,6 +53,9 @@ trait Logging[F[_]] {
   def trackException(e: Throwable): F[Unit]
+
+  /** Send metrics to StatsD if configured */
+  def sendMetrics(loaded: LoaderMessage.ShreddingComplete): F[Unit]
```
Also agree that it should be in `Logging` 👍

chuwy reviewed (changes requested):
Looks great!

The only blocker is human-readable logging - unless you have strong opinion why it should be deleted - we must bring it back.

chuwy commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/dsl/StatsD.scala:
```diff
@@ -0,0 +1,92 @@
+        loaded.timestamps.max.map(max => Duration.between(max, now).toSeconds()).map(l => KVMetric(MinCollectorLatencyName, l.toString)),
+        Some(Duration.between(loaded.timestamps.jobCompleted, now).toSeconds()).map(l => KVMetric(LoaderLatencyName, l.toString)),
+        loaded.count.map(c => KVMetric(CountGoodName, c.toString))
+      )
+        .collect { case Some(m) => m }
```
```suggestion
      ).unite
```

`unite` comes from [cats.Alternative](https://github.com/typelevel/cats/blob/main/core/src/main/scala/cats/Alternative.scala#L24-L27) and exists exactly for `List[Option[A]]` casts. `flatten` would also do the trick, but I don't like it because it implies `F[F[A]]` and works just because of Scala "hacks"

benjben replied:
That's great :ok_hand:

benjben commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/dsl/StatsD.scala:
```diff
@@ -0,0 +1,92 @@
+      List(
+        loaded.timestamps.min.map(min => Duration.between(min, now).toSeconds()).map(l => KVMetric(MaxCollectorLatencyName, l.toString)),
+        loaded.timestamps.max.map(max => Duration.between(max, now).toSeconds()).map(l => KVMetric(MinCollectorLatencyName, l.toString)),
+        Some(Duration.between(loaded.timestamps.jobCompleted, now).toSeconds()).map(l => KVMetric(LoaderLatencyName, l.toString)),
```
I think that both `jobStarted` and `jobCompleted` can bring value, I'll just send both.

benjben commented:
@chuwy I added the possibility to choose stdout to report the metrics, which required some changes in the design.
We now have `Reporter` that sends the list of metrics that it is given and `Metrics` that is in charge of just computing the metrics.
They are not aware of each other, `congratulates` [makes the link between them](https://github.com/snowplow/snowplow-rdb-loader/blob/feature/metrics/modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/loading/Load.scala#L129-L130).
I added a [unit test](https://github.com/snowplow/snowplow-rdb-loader/blob/feature/metrics/modules/loader/src/test/scala/com/snowplowanalytics/snowplow/rdbloader/dsl/metrics/MetricsSpec.scala) for the computaion of the metrics.

/cc @istreeter

istreeter commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/dsl/metrics/StdoutReporter.scala:
```diff
@@ -0,0 +1,31 @@
+    new Reporter[F] {
+      def report(metrics: List[Metrics.KVMetric]): F[Unit] = {
+        val formatted = metrics.map(formatMetric(stdoutConfig.prefix))
+        val metricsStr = s"METRICS - ${formatted.mkString(" ")}"
```
Do you think there is a need to make stdout formatting similar between Enrich and rdb loader?

Just FYI in Enrich the stdout metrics look like this

```
[ioapp-compute-1] INFO enrich.metrics - snowplow.enrich.raw = 20000
[ioapp-compute-1] INFO enrich.metrics - snowplow.enrich.good = 20000
[ioapp-compute-1] INFO enrich.metrics - snowplow.enrich.bad = 0
```

where...
- `[ioapp-compute-1] INFO enrich.metrics -` is the prefix added automatically by our slf4j configuration
- `enrich.metrics` is what we choose to call the slf4j logger
- `snowplow.enrich.raw = 20000` is what we send to the logger.
 
Note the spaces around the `=`, and one line per metric.

benjben replied:
Good point !

We don't use slf4jlogger in RDB loader so it won't be exactly the same but I will also use one line per metric with the ` = `.

benjben replied:
Actually I think that the time has come to add slf4jlogger to RDB loader

benjben commented:
- I updated the configuration format to be consistent with enrich and make it possible to use both StatsD and stdout for metrics
- I took the opportunity to make the unit tests use the example `config.hocon` and not redefine one in different places
- `slf4j` is now used for logging
- `track()` and `trackException()` got moved to `Monitoring`, with the new `reportMetrics()`

I haven't tested live yet

chuwy commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/dsl/metrics/StatsDReporter.scala:
```diff
@@ -0,0 +1,65 @@
+    statsDConfig match {
+      case Some(config) =>
+        new Reporter[F] {
+          def report(metrics: List[Metrics.KVMetric]): F[Unit] = {
+            mkSocket[F](blocker).use { socket =>
+              for {
+                formatted <- Sync[F].pure(metrics.map(statsDFormat(config)))
+                ip <- blocker.delay(InetAddress.getByName(config.hostname))
+                _ <- formatted.traverse_(sendMetric[F](blocker, socket, ip, config.port))
+              } yield ()
```
```suggestion
          def report(metrics: List[Metrics.KVMetric]): F[Unit] = {
            val formatted = metrics.map(statsDFormat(config))
            mkSocket[F](blocker).use { socket =>
              for {
                ip <- blocker.delay(InetAddress.getByName(config.hostname))
                _ <- formatted.traverse_(sendMetric[F](blocker, socket, ip, config.port))
              } yield ()
```

This is obviously super minor in this case, but generally you'd want to keep as few things inside `IO` as possible, especially if `ContextShift` is used.

chuwy reviewed (approved):
Hey @benjben! The PR looks really awesome! 

I think I haven't updated the `develop` branch when released 1.0.1, so I cherry-picked only statsd commit to review. But again - looks just perfect.

benjben commented on config/config.hocon.sample:
```diff
@@ -43,34 +44,39 @@
     "type": "redshift",
     # Redshift hostname
-    "host": "redshift.amazon.com",
+    "host": "angkor-wat-final.ccxvdpz01xnr.us-east-1.redshift.amazonaws.com",
```
That was my plan indeed. Do you think that I should use something more generic like `redshift.amazon.com` ?

chuwy replied:
Yup, I'd stick with something simple and generic, but it's by no means a strong preference.

benjben commented on build.sbt:
```diff
@@ -64,6 +64,7 @@ lazy val loader = project.in(file("modules/loader"))
     Compile / mainClass := Some("com.snowplowanalytics.snowplow.rdbloader.Main")
   )
   .settings(BuildSettings.buildSettings)
+  .settings(BuildSettings.addExampleConfToTestCp)
```
Yes but for `common` module, and now that loader also uses the example config I need it there as well

benjben commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/dsl/Environment.scala:
```diff
@@ -72,24 +73,26 @@ object Environment {
   }
 
-  def initSentry[F[_]: Sync](dsn: Option[URI]): Resource[F, Option[SentryClient]] =
+  def initSentry[F[_]: Sync](dsn: Option[URI], logging: Logging[F]): Resource[F, Option[SentryClient]] =
```
How can I provide `Logging` evidence in the `for` loop ?

I can't write something like that that:
```
logging = Logging.loggingInterpreter[F](List(cli.config.storage.password.getUnencrypted, cli.config.storage.username))
implicit loggingF: Logging[F] = logging
sentry <- initSentry[F](cli.config.monitoring.sentry.map(_.dsn))
```

I found [this ticket](https://github.com/scala/bug/issues/2823).

chuwy replied:
I didn't realise the reason and I think it's justified, but as it's mentioned in the ticket, you can do this with `better-monadic-for` plugin (I think we use it already): https://github.com/oleg-py/better-monadic-for#define-implicits-in-for-comprehensions-or-matches

benjben replied:
Nice!

benjben commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/dsl/Monitoring.scala:
```diff
@@ -0,0 +1,143 @@
+   * A sentry client, exposed as an impure member
+   * to handle the very final exception in Main class
+   */
+  def sentry: Option[SentryClient]
```
Indeed I could remove it
